### ISSUE TITLE
`for-of` loop is lost on Rollup 0.62.0

### ISSUE THREAD

**Issue by mysticatea** [rollup/rollup#2316]
Rollup deletes `for-of` loop in this simple code:

```js
export function f(list) {
  for (const el of list) {
    if (el != null) {
      el.foo = 0
    }
  }
}
```

> https://rollupjs.org/repl?version=0.62.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGZ1bmN0aW9uJTIwZihsaXN0KSUyMCU3QiU1Q24lNUN0Zm9yJTIwKGNvbnN0JTIwZWwlMjBvZiUyMGxpc3QpJTIwJTdCJTVDbiU1Q3QlNUN0aWYlMjAoZWwlMjAhJTNEJTIwbnVsbCklMjAlN0IlNUNuJTVDdCU1Q3QlNUN0ZWwuZm9vJTIwJTNEJTIwMCU1Q24lNUN0JTVDdCU3RCU1Q24lNUN0JTdEJTVDbiU3RCU1Q24lMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==

**Comment by mysticatea** [comment]
I confirmed that `0.60.x` has no problem. It seemed to introduce since `0.61`.

**Comment by lukastaegert** [comment]
Thanks a lot! The issue is actually related to the literal value logic that is used to resolve the if-statement. I.e. rollup thinks that `el` is always `undefined`. No idea what I was thinking when writing this, for some reason the logic was only doing this properly when the loop variable was not declared as part of the loop. I added a fix to #2315.

Btw. this issue extends to for-in statements symmetrically:

https://rollupjs.org/repl?version=0.62.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmNvbnN0JTIwbGlzdCUyMCUzRCUyMCU1QiU3QiU3RCU1RCUzQiU1Q24lNUNuZm9yJTIwKGNvbnN0JTIwZWxlbWVudCUyMG9mJTIwbGlzdCklMjAlN0IlNUNuJTVDdGlmJTIwKGVsZW1lbnQpJTIwJTdCJTVDbiU1Q3QlNUN0ZWxlbWVudC5mb28lMjAlM0QlMjB0cnVlJTNCJTVDbiU1Q3QlN0QlNUNuJTdEJTVDbiU1Q25hc3NlcnQub2sobGlzdCU1QjAlNUQuZm9vJTJDJTIwJ0xpc3QlMjBlbGVtZW50JTIwd2FzJTIwbm90JTIwcmVhc3NpZ25lZCcpJTNCJTVDbiU1Q25jb25zdCUyMG9iaiUyMCUzRCUyMCU3QiU1Q24lNUN0eCUzQSUyMGZhbHNlJTVDbiU3RCUzQiU1Q24lNUNuZm9yJTIwKGNvbnN0JTIwa2V5JTIwaW4lMjBvYmopJTIwJTdCJTVDbiU1Q3RpZiUyMChrZXkpJTIwJTdCJTVDbiU1Q3QlNUN0b2JqJTVCa2V5JTVEJTIwJTNEJTIwdHJ1ZSUzQiU1Q24lNUN0JTdEJTVDbiU3RCU1Q24lNUNuYXNzZXJ0Lm9rKG9iai54JTJDJTIwJ09iamVjdCUyMHByb3BlcnR5JTIwd2FzJTIwbm90JTIwcmVhc3NpZ25lZCcpJTNCJTVDbiU1Q24lMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==

**Comment by lukastaegert** [comment]
Fix released via Rollup v0.63.0, thanks for being patient!
